### PR TITLE
feat: [WIP] Add audit trail for entire event  [DHIS2-11481]

### PR DESCRIPTION
**Pull request by zubaira**
PR will add support for the audit trail covering the entire event (ProgramStageInstance).

- [x] Audit trail object
- [x] Service layer and repository 
- [x] Endpoint in AuditController

### CODE REVIEW

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AuditController.java [hunk 1]
```diff
@@ -149,6 +161,8 @@ public AuditController( IdentifiableObjectManager manager, ProgramStageInstanceS
         this.fieldFilterService = fieldFilterService;
         this.contextService = contextService;
         this.fileResourceService = fileResourceService;
+        this.programStageInstanceAuditService = programStageInstanceAuditService;
+        this.programStageService = programStageService;
```
Comment (enricocolasante) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AuditController.java:165]: Is ProgramStageService used anywhere?

File: dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.37/V2_37_22__Create_table_for_program_stage_instance_audit.sql [hunk 2]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE programstageinstanceaudit (
```
Comment (enricocolasante) [inline dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.37/V2_37_22__Create_table_for_program_stage_instance_audit.sql:1]: It is a good practice to make all the scripts idempotent.
`CREATE TABLE IF EXISTS` will do the trick

**Review by enricocolasante — CHANGES_REQUESTED** [review]
A couple of comments

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AuditController.java (hunk 1 above)
Comment (zubaira) [reply dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AuditController.java:165]: resolved

File: dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.37/V2_37_22__Create_table_for_program_stage_instance_audit.sql (hunk 2 above)
Comment (zubaira) [reply dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.37/V2_37_22__Create_table_for_program_stage_instance_audit.sql:1]: resolved

**Review by enricocolasante — APPROVED** [review]

**Comment by zubaira** [comment]
The exiting audit framework was partially providing this functionality due to some bugs. Those bugs are not fixed so we don't need this fix.

**Pull request closed without merging**